mainnet-013: Fault Proofs Fjord Upgrade (#232)
ethereum-optimism/superchain-ops: tasks/eth/013-fp-upgrade-fjord/VALIDATION.md

      - The location (`8`) of the `approvedHashes` mapping in the [GnosisSafe storage layout](https://github.com/safe-global/safe-contracts/blob/v1.3.0/contracts/examples/libraries/GnosisSafeStorage.sol#L20)
      - The address of the Council Safe: `0xc2819DC788505Aac350142A7A707BF9D03E3Bd03`
      - The safe hash to approve: `0x530085da61c08016f1625301df4686fcfe40291b98db29a4bf801237ca3098c4`
    - The using `cast index`, we can verify that:
      ```shell
        $ cast index address 0xc2819DC788505Aac350142A7A707BF9D03E3Bd03 8
        0xaaf2b641eaf0bae063c4f2e5670f905e1fb7334436b902d1d880b05bd6228fbd
        ```
        and
      ```shell
        $ cast index bytes32 0x530085da61c08016f1625301df4686fcfe40291b98db29a4bf801237ca3098c4 0xaaf2b641eaf0bae063c4f2e5670f905e1fb7334436b902d1d880b05bd6228fbd
        0x0f7b79f6b38abe5e02d33eec6fcdf7d9447ff17f6803d46f0afd5f628bac8504 
        ```
      And so the output of the second command matches the key above.

- **Key:** `0x24424336f04440b1c28685a38303ac33c9d14a25`
  **Before:** `0x0000000000000000000000000000000000000000000000000000000000000000`
  **After:** `0x000000000000000000000000000000000000000000000000000000006675b61f`

>   **Meaning:** This updates the [`lastLive`](https://github.com/ethereum-optimism/optimism/blob/op-contracts/v1.5.0/packages/contracts-bedrock/src/Safe/LivenessGuard.sol#L36) indicating liveness of an owner that participated in signing. This will be updated to a block timestamp that's close to when this task was executed.

Reviewer: I think you are missing the header here since this is the Liveness Guard address, not the 2/2 Proxy Admin Owner. Also, the before value may not be zero for some signers—we can just note that in the Meaning text

Author: Good catch. Thanks for your patience 😅

# VALIDATION

This document can be used to validate the state diff resulting from the execution of the upgrade
transaction.

For each contract listed in the state diff, please verify that no contracts or state changes shown in the Tenderly diff are missing from this document. Additionally, please verify that for each contract:

- The following state changes (and none others) are made to that contract. This validates that no unexpected state changes occur.
- All addresses (in section headers and storage values) match the provided name, using the Etherscan and Superchain Registry links provided. This validates the bytecode deployed at the addresses contains the correct logic.
- All key values match the semantic meaning provided, which can be validated using the storage layout links provided.

## State Overrides

The following state overrides should be seen:

> ### `0x5a0Aae59D09fccBdDb6C6CcEB07B7279367C3d2A` (The 2/2 `ProxyAdmin` Owner)

Reviewer: Can we sort the address keys here to match the order they are shown in tenderly (alphabetical)? This just makes it easier for singers to cross-check everything

  **Meaning:** The nonce is increased from 5 to 6. The key can be validated by the location of the `nonce` variable in the [Safe's Storage Layout](https://github.com/safe-global/safe-smart-account/blob/v1.3.0/contracts/examples/libraries/GnosisSafeStorage.sol#L17).

### `0x847b5c174615b1b7fdf770882256e2d3e95b9d92` (Foundation Safe)

State Changes:

#### For the Foundation:

- **Key:** `0x0000000000000000000000000000000000000000000000000000000000000005` <br/>
  **Before:** `0x0000000000000000000000000000000000000000000000000000000000000005` <br/>
  **After:** `0x0000000000000000000000000000000000000000000000000000000000000006` <br/>
  **Meaning:** The nonce is increased from 5 to 6. The key can be validated by the location of the `nonce` variable in the [Safe's Storage Layout](https://github.com/safe-global/safe-smart-account/blob/v1.3.0/contracts/examples/libraries/GnosisSafeStorage.sol#L17). Note that if this task is simulated before 011 has executed, then the nonce should be increased from 4 to 5.

### `0x24424336F04440b1c28685a38303aC33C9D14a25` (`LivenessGuard`)

State Changes:

#### For the Council:

> - **Key:** `0xee4378be6a15d4c71cb07a5a47d8ddc4aba235142e05cb828bb7141206657e27` <br/>

Reviewer: ```suggestion
- **Key:** Compute with `cast index address {yourSignerAddressHere} 0` <br/>
```

  **Before:** `0x0000000000000000000000000000000000000000000000000000000000000000` <br/>
  **After:** `0x000000000000000000000000000000000000000000000000000000006675b61f` <br/>

>   **Meaning:** This updates the [`lastLive`](https://github.com/ethereum-optimism/optimism/blob/op-contracts/v1.5.0/packages/contracts-bedrock/src/Safe/LivenessGuard.sol#L36) indicating liveness of an owner that participated in signing. This will be updated to a block timestamp that's close to when this task was executed. Note that the "before" value may be non-zero for signers that have participated in signing.

Reviewer: ```suggestion
  **Meaning:** This updates the [`lastLive`](https://github.com/ethereum-optimism/optimism/blob/op-contracts/v1.5.0/packages/contracts-bedrock/src/Safe/LivenessGuard.sol#L36) indicating liveness of an owner that participated in signing. This will be updated to a block timestamp that matches the time when this task was executed. Note that the "before" value may be non-zero for signers that have participated in signing.
  
The only other state change is a nonce increment of the account that sent the transaction.
```

This mapping implements a linked list for iterating through the list of owners. Since we'll only have one owner (Multicall), and the `0x01` address is used as the first and last entry in the linked list, we will see the following overrides:
- `owners[1] -> 0xca11bde05977b3631167028862be2a173976ca11`
- `owners[0xca11bde05977b3631167028862be2a173976ca11] -> 1`

And we do indeed see these entries:

- **Key:** 0x316a0aac0d94f5824f0b66f5bbe94a8c360a17699a1d3a233aafcf7146e9f11c <br/>
  **Value:** 0x0000000000000000000000000000000000000000000000000000000000000001 <br/>
  **Meaning:** This is `owners[0xca11bde05977b3631167028862be2a173976ca11] -> 1`, so the key can be
    derived from `cast index address 0xca11bde05977b3631167028862be2a173976ca11 2`.

- **Key:** 0xe90b7bceb6e7df5418fb78d8ee546e97c83a08bbccc01a0644d599ccd2a7c2e0 <br/>
  **Value:** 0x000000000000000000000000ca11bde05977b3631167028862be2a173976ca11 <br/>
  **Meaning:** This is `owners[1] -> 0xca11bde05977b3631167028862be2a173976ca11`, so the key can be
    derived from `cast index address 0x0000000000000000000000000000000000000001 2`.

## State Changes

> ### `0x24424336F04440b1c28685a38303aC33C9D14a25` (`LivenessGuard`)
> 
> State Changes:
> 
> #### For the Council:
> 
> - **Key:** Compute with `cast index address {yourSignerAddress} 0` <br/>

Reviewer: nit, just a little more readable 
```suggestion
### `0x24424336F04440b1c28685a38303aC33C9D14a25` (`LivenessGuard`, Council only)

State Changes:

- **Key:** Compute with `cast index address {yourSignerAddress} 0` <br/>
```

- **Key:** `0x4d5a9bd2e41301728d41c8e705190becb4e74abe869f75bdb405b63716a35f9e` <br/>
  **Before:** `0x000000000000000000000000e9dad167ef4de8812c1abd013ac9570c616599a0` <br/>
  **After:** [`0x000000000000000000000000c307e93a7c530a184c98eade4545a412b857b62f`](https://etherscan.io/address/0xc307e93a7c530a184c98eade4545a412b857b62f) <br/>
  **Meaning:** This is `gameImpls[0] -> 0xc307e93a7c530a184c98eade4545a412b857b62f` (where `0` is the [`CANNON` game type](https://github.com/ethereum-optimism/optimism/blob/op-contracts/v1.4.0/packages/contracts-bedrock/src/dispute/lib/Types.sol#L28), so the key can be derived from `cast index uint32 0 101`.

- **Key:** `0xffdfc1249c027f9191656349feb0761381bb32c9f557e01f419fd08754bf5a1b` <br/>
  **Before:** `0x0000000000000000000000004146df64d83acb0dcb0c1a4884a16f090165e122` <br/>
  **After:** [`0x000000000000000000000000f691f8a6d908b58c534b624cf16495b491e633ba`](https://etherscan.io/address/0xf691f8a6d908b58c534b624cf16495b491e633ba) <br/>

>   **Meaning:** This is `gameImpls[1] -> 0xf691f8a6d908b58c534b624cf16495b491e633ba` (where `1` is the [`PERMISSIONED_CANNON` game type](https://github.com/ethereum-optimism/optimism/blob/op-contracts/v1.4.0/packages/contracts-bedrock/src/dispute/lib/Types.sol#L31)), so the key can be derived from `cast index uint32 0 101`.

Reviewer: ```suggestion
  **Meaning:** This is `gameImpls[1] -> 0xf691f8a6d908b58c534b624cf16495b491e633ba` (where `1` is the [`PERMISSIONED_CANNON` game type](https://github.com/ethereum-optimism/optimism/blob/op-contracts/v1.4.0/packages/contracts-bedrock/src/dispute/lib/Types.sol#L31)), so the key can be derived from `cast index uint32 0 101`.
  
The only other state changes is a nonce increments on the account that sent the transaction.
```

This mapping implements a linked list for iterating through the list of owners. Since we'll only have one owner (Multicall), and the `0x01` address is used as the first and last entry in the linked list, we will see the following overrides:
- `owners[1] -> 0xca11bde05977b3631167028862be2a173976ca11`
- `owners[0xca11bde05977b3631167028862be2a173976ca11] -> 1`

And we do indeed see these entries:

- **Key:** 0x316a0aac0d94f5824f0b66f5bbe94a8c360a17699a1d3a233aafcf7146e9f11c <br/>
  **Value:** 0x0000000000000000000000000000000000000000000000000000000000000001 <br/>
  **Meaning:** This is `owners[0xca11bde05977b3631167028862be2a173976ca11] -> 1`, so the key can be
    derived from `cast index address 0xca11bde05977b3631167028862be2a173976ca11 2`.

- **Key:** 0xe90b7bceb6e7df5418fb78d8ee546e97c83a08bbccc01a0644d599ccd2a7c2e0 <br/>
  **Value:** 0x000000000000000000000000ca11bde05977b3631167028862be2a173976ca11 <br/>
  **Meaning:** This is `owners[1] -> 0xca11bde05977b3631167028862be2a173976ca11`, so the key can be
    derived from `cast index address 0x0000000000000000000000000000000000000001 2`.

> ## State Changes

Reviewer: Needs an entry for state changes in the LivenessGuard, for Security Council only. The mapping key used for the slot calculation is the signer's own address, or the multicall3 address when simulating without ledger. The value written will be the timestamp of the simulation (or execution) so it will be different for everyone

Author: Confirming that this task, and neither 011 nor 012, will be the first to update the liveness check. So we expect the initial timestamp at the time of execution to be zeroed.

- **Key:** `0xffdfc1249c027f9191656349feb0761381bb32c9f557e01f419fd08754bf5a1b` <br/>
  **Before:** `0x0000000000000000000000004146df64d83acb0dcb0c1a4884a16f090165e122` <br/>
  **After:** [`0x000000000000000000000000f691f8a6d908b58c534b624cf16495b491e633ba`](https://etherscan.io/address/0xf691f8a6d908b58c534b624cf16495b491e633ba) <br/>
  **Meaning:** This is `gameImpls[1] -> 0xf691f8a6d908b58c534b624cf16495b491e633ba` (where `1` is the [`PERMISSIONED_CANNON` game type](https://github.com/ethereum-optimism/optimism/blob/op-contracts/v1.4.0/packages/contracts-bedrock/src/dispute/lib/Types.sol#L31)), so the key can be derived from `cast index uint32 0 101`.

### Safe Contract State Changes

The only other state changes should be restricted to one of the following addresses:

- L1 2/2 ProxyAdmin Owner Safe: `0x5a0Aae59D09fccBdDb6C6CcEB07B7279367C3d2A`
  - The nonce (slot 0x5) should be increased from 4 to 5.

>   - Another key is set from 0 to 1 reflecting an entry in the `approvedHashes` mapping.

Reviewer: We need to explain how to verify that this key is correct. Let's pull this out into a ful `### 0x5a0Aae59D09fccBdDb6C6CcEB07B7279367C3d2A` state changes entry. Here is a reference implementation, you will just need to change the keys and the safe hashes: https://github.com/ethereum-optimism/superchain-ops/blob/main/tasks/eth/010-1-guardian-upgrade/VALIDATION.md#0x5a0aae59d09fccbddb6c6cceb07b7279367c3d2a-the-22-proxyadmin-owner-1

Author: That example was very useful. Do we also want to specify the expected nonce changes for the SC and FND safes? Or is that redundant given that the approval hash includes it?

Reviewer: I think it's good to specify, it gives extra confidence to signers that they are signing the right thing, since it's more user-friendly than the approveHash
